cmd/clusterctl/client/alpha/rollout_statuswatcher.go

```diff
@@ -0,0 +1,113 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alpha
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+)
+
+// StatusWatcher provides an interface for resources that have rollout status.
+type StatusWatcher interface {
+	Status(ctx context.Context, proxy cluster.Proxy, name, namespace string) (string, bool, error)
+}
+
+// KubeadmControlPlaneStatusWatcher implements the StatusViewer interface.
+type KubeadmControlPlaneStatusWatcher struct{}
+
+// MachineDeploymentStatusWatcher implements the StatusViewer interface.
+type MachineDeploymentStatusWatcher struct{}
+
+// ObjectStatusWatcher will issue a view on the specified cluster-api resource.
+func (r *rollout) ObjectStatusWatcher(ctx context.Context, proxy cluster.Proxy, ref corev1.ObjectReference) error {
+	log := ctrl.LoggerFrom(ctx)
+	var statusViewer StatusWatcher
+	switch ref.Kind {
+	case KubeadmControlPlane:
+		statusViewer = &KubeadmControlPlaneStatusWatcher{}
+	case MachineDeployment:
+		statusViewer = &MachineDeploymentStatusWatcher{}
+	default:
+		return errors.Errorf("invalid resource type %q, valid values are %v", ref.Kind, validResourceTypes)
+	}
+
+	oldStatus := ""
+	timeout := 600 * time.Second
+	lastChange := time.Now()
+	for {
+		status, done, err := statusViewer.Status(ctx, proxy, ref.Name, ref.Namespace)
+		if err != nil {
+			log.Error(err, "failed to get resource status")
+			time.Sleep(2 * time.Second)
+			continue
+		}
+		if status != oldStatus {
+			fmt.Fprintf(os.Stdout, "%s", status)
+			oldStatus = status
+			lastChange = time.Now()
+		}
+		if done || time.Since(lastChange) > timeout {
+			break
+		}
+		time.Sleep(2 * time.Second)
+	}
+	return nil
+}
+
+// Status returns a status describing kubeadmcontrolplane status, and a bool value indicating if the status is considered done.
+func (s *KubeadmControlPlaneStatusWatcher) Status(ctx context.Context, proxy cluster.Proxy, name, namespace string) (string, bool, error) {
+	newKcp, err := getKubeadmControlPlane(ctx, proxy, name, namespace)
+	if err != nil || newKcp == nil {
+		return "", false, errors.Wrapf(err, "failed to get kubeadmcontrolplane/%v", name)
+	}
+	if newKcp.Spec.Replicas != nil && newKcp.Status.UpdatedReplicas < *newKcp.Spec.Replicas {
+		return fmt.Sprintf("Waiting for kubeadmcontrolplane %q rollout to finish: %d out of %d new replicas have been updated...\n", newKcp.Name, newKcp.Status.UpdatedReplicas, *newKcp.Spec.Replicas), false, nil
```

Comment:
Q: Is there a way to reuse something like the ready condition in the status? That should already provide a summary of the resource as well as info on if it's an error, warning, or informational message.